src/number_theory/zeta_values.lean

```diff
@@ -0,0 +1,137 @@
+Authors: David Loeffler
+-/
+
+import number_theory.bernoulli_polynomials
+import analysis.special_functions.integrals
+import measure_theory.integral.interval_integral
+import analysis.fourier.add_circle
+
+/-!
+# Critical values of the Riemann zeta function
+
+In this file we evaluate the Fourier coefficients of Bernoulli polynomials on the interval `[0, 1]`.
+In a future iteration this will be used to evaluate critical values of the Riemann zeta function
+(and other Dirichlet L-functions).
+-/
+
+noncomputable theory
+open_locale nat real interval
+open complex measure_theory set interval_integral
+
+section bernoulli_fun_props
+/-! Simple properties of the Bernoulli polynomial, as a function `ℝ → ℝ`. -/
+
+/-- The function `x ↦ Bₖ(x) : ℝ → ℝ`. -/
+def bernoulli_fun (k : ℕ) (x : ℝ) : ℝ :=
+(polynomial.map (algebra_map ℚ ℝ) (polynomial.bernoulli k)).eval x
+
+lemma bernoulli_fun_eval_zero (k : ℕ) : bernoulli_fun k 0 = bernoulli k :=
+by rw [bernoulli_fun, polynomial.eval_zero_map, polynomial.bernoulli_eval_zero, eq_rat_cast]
+
+lemma bernoulli_fun_endpoints_eq_of_ne_one {k : ℕ} (hk : k ≠ 1) :
+  bernoulli_fun k 1 = bernoulli_fun k 0 :=
+by rw [bernoulli_fun_eval_zero, bernoulli_fun, polynomial.eval_one_map,
+  polynomial.bernoulli_eval_one, bernoulli_eq_bernoulli'_of_ne_one hk, eq_rat_cast]
+
+lemma bernoulli_fun_eval_one (k : ℕ) : bernoulli_fun k 1 = bernoulli_fun k 0 + ite (k = 1) 1 0 :=
+begin
+  rw [bernoulli_fun, bernoulli_fun_eval_zero, polynomial.eval_one_map,
+    polynomial.bernoulli_eval_one],
+  split_ifs,
+  { rw [h, bernoulli_one, bernoulli'_one, eq_rat_cast],
+    push_cast, ring },
+  { rw [bernoulli_eq_bernoulli'_of_ne_one h, add_zero, eq_rat_cast], }
+end
+
+lemma has_deriv_at_bernoulli_fun (k : ℕ) (x : ℝ) :
+  has_deriv_at (bernoulli_fun k) (k * bernoulli_fun (k - 1) x) x :=
+begin
+  convert ((polynomial.bernoulli k).map $ algebra_map ℚ ℝ).has_deriv_at x using 1,
+  simp only [bernoulli_fun, polynomial.derivative_map, polynomial.derivative_bernoulli k,
+    polynomial.map_mul, polynomial.map_nat_cast, polynomial.eval_mul, polynomial.eval_nat_cast],
+end
+
+lemma antideriv_bernoulli_fun (k : ℕ) (x : ℝ) :
+  has_deriv_at (λ x, (bernoulli_fun (k + 1) x) / (k + 1)) (bernoulli_fun k x) x :=
+begin
+  convert (has_deriv_at_bernoulli_fun (k + 1) x).div_const _,
+  field_simp [nat.cast_add_one_ne_zero k],
+  ring,
+end
+
+lemma integral_bernoulli_fun_eq_zero (k : ℕ) (hk : 1 ≤ k) :
+  ∫ (x : ℝ) in 0..1, bernoulli_fun k x = 0 :=
+begin
+  rw integral_eq_sub_of_has_deriv_at (λ x hx, antideriv_bernoulli_fun k x)
+    ((polynomial.continuous _).interval_integrable _ _),
+  dsimp only,
+  rw bernoulli_fun_eval_one,
+  split_ifs,
+  { exfalso, linarith, }, { simp },
+end
+
+end bernoulli_fun_props
+
+section bernoulli_fourier_coeffs
+/-! Compute the Fourier coefficients of the Bernoulli functions via integration by parts. -/
+
+private def hT := fact.mk (zero_lt_one' ℝ)
+local attribute [instance] hT
```

Comment:
There are many private lemmas called `fact_zero_lt_one` scattered around the library. Could you put one nonprivate version in an early file, for instance in `real.lean`, and use it everywhere (including here) with `local attribute [instance] real.fact_zero_lt_one`?

Reply:
I'm wary of opening this can of worms. This isn't specific to the real numbers, so arguably it should go in `src/algebra/order/zero_le_one.lean`, but that file is already synchronised to mathlib4. 

I'd also argue that the correct solution is different. All but one of the uses of `fact_zero_lt_one` are inherited from `add_circle`.  But the only properties of `add_circle T` which are still valid for `T = 0` are uninteresting anyway, since `add_circle (0:k) = k`. So I'd suggest that `add_circle` should take a proof of `0 < T` as an input to its constructor; then we can avoid all this `fact` business entirely.

Reply:
In any case, I'd suggest that this should be fixed in a different PR rather than this one.

Reply:
The fact is used several times in the library (not only for `add_circle`), always over the reals, with 4 competing lemmas with exactly the same content. It's just more hygienic to merge these 4 lemmas into one. And it should only be over the reals, because `zero_lt_one` has sometimes trouble figuring out the right type, so better only have the useful case in our lemma. Anyway, I have pushed the cleanup, let's see what CI thinks of it.

Reply:
> The fact is used several times in the library (not only for add_circle), always over the reals, with 4 competing lemmas with exactly the same content.

Three of those four are only needed in order to use `add_circle`; the sole exception is `geometry.manifold.instances.real`. But I'm happy to do this whichever way you prefer.